### PR TITLE
Conditions (WIP)

### PR DESCRIPTION
**Pull request by andymoody**
# Description

* Break inline conditions amendment into separate components. Also offer the user the ability to select an existing condition.
* Some minor amends around ui (cancel and edit link rationalisation)

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce.

- [X] Unit tests
- [X] Click through testing via page-create.js (Simply revert my last commit to do this)

# Checklist:

- [X] My code follows the [javascript standard style](https://standardjs.com/)
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have made corresponding changes to the documentation and versioning
- [X] My changes generate no new warnings
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes
- [X] I have rebased onto master and there are no code conflicts

### CODE REVIEW

File: designer/client/conditions/inline-condition-model.js [hunk 1]
```diff
@@ -21,6 +21,13 @@ export class ConditionsModel {
...
   }
 
+  clear () {
+    this.#userGroupedConditions = []
+    this.#groupedConditions = []
+    this.#conditionName = undefined
+    return this
+  }
+
   name (name) {
```
Comment (jenbutongit) [inline designer/client/conditions/inline-condition-model.js:31]: should this be a setter? (and getter). name looks like it just generates a name, but it's also doing some mutations. Feel free to leave as is, maybe just add a jsdoc?

File: designer/client/conditions/inline-conditions-definition.js [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+    if (input.value && input.value.trim() !== '') {
+      const copy = clone(this.state.condition??{})
+      copy.coordinator = input.value
+      this.setState({
+        condition: copy
+      })
+    } else {
+      this.setState({
+        condition: {}
+      })
```
Comment (jenbutongit) [inline designer/client/conditions/inline-conditions-definition.js:50]: can we do just one setState? `let update = {}`, set update inside the if block, then `this.setState(...condition: update)`. that way we can get rid of else

**Comment by jenbutongit** [comment]
Looks good so far!

File: designer/client/conditions/inline-condition-model.js (hunk 1 above)
Comment (andymoody) [reply designer/client/conditions/inline-condition-model.js:31]: Fair point, I'll amend with my work on getting it wired in properly and persisted

File: designer/client/conditions/inline-conditions-definition.js (hunk 2 above)
Comment (andymoody) [reply designer/client/conditions/inline-conditions-definition.js:50]: ok

File: designer/client/conditions/inline-conditions.js [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+    })
+  }
+
+  editCallback = (conditions) => {
+    this.setState({
+      conditions: conditions
+    })
+  }
+
+  exitEdit = () => {
```
[the remark this answers is not in the document]
Comment (andymoody) [reply designer/client/conditions/inline-conditions.js:102]: Done

File: designer/client/conditions/inline-conditions-edit.js [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+      selectedConditions: []
+    })
+  }
+
+  onClickMoveLater (index) {
+    this.setState({
+      conditions: this.state.conditions.moveLater(index),
+      selectedConditions: []
+    })
+  }
```
[the remark this answers is not in the document]
Comment (andymoody) [reply designer/client/conditions/inline-conditions-edit.js:213]: Done

**Comment by andymoody** [comment]
Some things I've left purposefully untouched and the reasons why:

* Condition selection on list items - I think we need to look at the UI for this before we can integrate inline condition building.
* Condition selection on component creation - I think we plan to do some work to support conditions across all component types, I thought integrating inline condition building here was best done once that is completed.
* Editing of conditions - I think this will be simple enough but thought it ought to come under a separate ticket.
* Additional operators - I think these are probably best driven by user feedback, unless you're aware of any that are obviously missing

**Pull request merged**